text/0010-bft-signatures.md

```diff
@@ -0,0 +1,291 @@
+    bytes nonce = 4;
+}
+```
+
+When a Fabric node validates a signature on a block, an ephemeral `SignatureHeader` is constructed where the `nonce` is taken from the
+`MetadataSignature` and the `creator` is the `identity` matching the `signer_id`. 
+
+This technique reduces the aforementioned overhead of 3.5KB to a meager 40 bytes.
+
+
+While the approach of the SmartBFT library integration works, it creates a coupling between the consensus implementation and the rest of Fabric, and this limits incorporation of other consensus implementations into Fabric.
+ 
+More specifically, it requires peers to understand how to decode the consensus metadata of ordering service nodes. That means that the peer needs to be aware of the type of ordering service it is working with, which breaks the abstraction of the ordering service as a service that receives as input transactions, and outputs blocks.
+
+
+# Proposed techniques
+[technique]: #proposed-technique
+
+The proposed approach is heavily inspired by SmartBFT but is designed to be consensus agnostic, minimize implementation and operational complexity.
+
+### Block replication
+
+The proposed technique for block replication is essentially the approach of SmartBFT where a re-shuffling takes place not only upon suspicion of the block source but also in a timely fashion in order to be resilient to overloaded or deliberately slow orderers.   
+In order to save network bandwidth and processing power when the network is idle, the automatic time-based re-shuffling will only take place if new blocks are received.
+Moreover, there will not be a **need** to specify in the peer's local configuration (or an orderer's local configuration) which of the implementations to use. 
+Notice that a block validation policy that mandates signatures from multiple ordering nodes, is **never** satisfied by **only** principal sets of size one. 
+Or, equivalently, a block validation policy that is satisfied by **only** principal sets of size one never requires signatures from multiple ordering nodes. 
+Evidently, policies of the former type are **always** BFT policies, while policies of the latter type might be either a crash fault policy or a BFT policy that utilizes a threshold signature scheme.
+
+Therefore, Fabric nodes will pick the BFT implementation if either the block validation policy implies the need, or if an override is activated in the node's local configuration.  
+
+
+### Block signature verification
+
+The proposed technique for block signature verification should be generic and consensus agnostic, in order to facilitate incorporation of multiple consensus types without the need for Fabric nodes to understand which consensus implementation is employed.
+
+Apart from the already existing block validation policy:
+
+```
+# BlockValidation specifies what signatures must be included in the block
+# from the orderer for the peer to validate it.
+BlockValidation:
+    Type: ImplicitMeta
+    Rule: "ANY Writers"
+```
+
```

Comment:
Can you provide an example of BlockValidation policy in the case of SmartBFT?

Reply:
So in the original SmartBFT, the block validation policy was actually computed dynamically ([implicitly](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/common/policies.proto#L16-L27)).  In short - the node would look at the last configuration block, and compose a policy that is made of [identity](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/msp/msp_principal.proto#L45) principals according to the [identities](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/orderer/smartbft/configuration.proto#L25) of the consenters and would require a [threshold](https://github.com/SmartBFT-Go/fabric-protos/blob/release-2.3-BFT/common/policies.proto#L42-L43) of `2f+1` signatures for the policy to be satisfied. 

In the proposed design, the policy can simply be a regular signature policy (not an implicit one!) that will have identities in it. However, when signatures will be encoded, they will be encoded using the references of the consenters. Then, when the signature is to be evaluated, the reference will be used to look up the identity (from the consenter mapping) and then the regular policy engine will be activated, over the regular signature policy in a manner that is transparent to the signature policy engine.